misc/futures-bounded/src/unique.rs

```diff
@@ -0,0 +1,230 @@
+    inner: HashMap<ID, (BoxFuture<'static, O>, Delay)>,
+    empty_waker: Option<Waker>,
+    full_waker: Option<Waker>,
+}
+
+/// Error of a worker pushing
+#[derive(PartialEq, Debug)]
+pub enum PushError<F> {
+    /// The length of the set is equal to the capacity
+    BeyondCapacity(F),
+    /// The set already contains the given worker's ID
+    ReplacedWorker(F),
+}
+
+impl<ID, O> UniqueWorkers<ID, O> {
+    pub fn new(timeout: Duration, capacity: usize) -> Self {
+        Self {
+            timeout,
+            capacity,
+            inner: Default::default(),
+            empty_waker: None,
+            full_waker: None,
+        }
+    }
+}
+
+impl<ID, O> UniqueWorkers<ID, O>
+where
+    ID: Clone + Hash + Eq + Send + 'static,
+{
+    /// Push a worker into the set.
+    /// This method adds the given worker with defined `worker_id` to the set.
+    /// If the length of the set is equal to the capacity, this method returns [PushError::BeyondCapacity],
+    /// that contains the passed worker. In that case, the worker is not added to the set.
+    /// If a worker with the given `worker_id` already exists, then the old worker will be replaced by a new one.
+    /// In that case, the returned error [PushError::ReplacedWorker] contains the old worker.
+    pub fn try_push<F>(&mut self, worker_id: ID, worker: F) -> Result<(), PushError<BoxFuture<O>>>
+    where
+        F: Future<Output = O> + Send + 'static + Unpin,
+    {
+        if self.inner.len() >= self.capacity {
+            return Err(PushError::BeyondCapacity(worker.boxed()));
+        }
+
+        let val = self.inner.insert(
+            worker_id.clone(),
+            (worker.boxed(), Delay::new(self.timeout)),
+        );
+
+        if let Some(waker) = self.empty_waker.take() {
+            waker.wake();
+        }
+
+        match val {
+            Some((old_worker, _)) => Err(PushError::ReplacedWorker(old_worker)),
+            None => Ok(()),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    pub fn poll_ready_unpin(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        if self.inner.len() < self.capacity {
+            return Poll::Ready(());
+        }
+
+        self.full_waker = Some(cx.waker().clone());
+
+        Poll::Pending
+    }
+
+    pub fn poll_unpin(&mut self, cx: &mut Context<'_>) -> Poll<(ID, Result<O, Timeout>)> {
+        let res = self
+            .inner
+            .iter_mut()
+            .find_map(|(worker_id, (worker, timeout))| {
+                if timeout.poll_unpin(cx).is_ready() {
+                    return Some((worker_id.clone(), Err(Timeout::new())));
+                }
+
+                match worker.poll_unpin(cx) {
+                    Poll::Ready(output) => Some((worker_id.clone(), Ok(output))),
+                    Poll::Pending => None,
+                }
+            });
+
+        match res {
+            None => {
+                self.empty_waker = Some(cx.waker().clone());
```

Comment:
Good point!
I suggest making an explicit check for empty map.

Reply:
I've been thinking, there are no tests to prove the correctness of working `empty_waker` and `full_waker. 
It looks wrong, isn't?

Reply:
I pushed a commit that uses `FuturesUnordered` under the hood which should be more efficient because it polls only the futures that actually need to be polled.